### PR TITLE
stop using `__sync_synchronize` on the hot path

### PR DESCRIPTION
**Pull request by kazuho**
Perhaps we can assume C compilers to have the atomic intrinsics that borrow the semantics from C++11? Doing so would help us optimize the hot path without making the code complex.

~~ToDo: discuss if we should retain the old code (for compatibility reasons), and if we decide to do so, how to switch between the two implementations.~~ We can consider using a macro to support the old way, if somebody complains.

### CODE REVIEW

File: include/h2o/multithread.h [hunk 1]
```diff
@@ -113,17 +113,16 @@ h2o_loop_t *h2o_multithread_get_loop(h2o_multithread_queue_t *);
...
             pthread_mutex_lock(&mutex);                                                                                            \
             if (!lock) {                                                                                                           \
                 do {                                                                                                               \
                     block                                                                                                          \
                 } while (0);                                                                                                       \
-                __sync_synchronize();                                                                                              \
-                lock = 1;                                                                                                          \
+                __atomic_store_n(&lock, 1, __ATOMIC_RELEASE);                                                                        \
```
Comment (robguima) [inline include/h2o/multithread.h:125]: cosmetic: backslash is not aligned.

**Review by robguima — APPROVED** [review]
lgtm thank you!

File: include/h2o/multithread.h (hunk 1 above)
Comment (kazuho) [reply include/h2o/multithread.h]: Nice catch! Fixed in 7305746.

**Comment by kazuho** [comment]
Thank you for checking. To be merge once CI concludes success.

**Pull request merged**